### PR TITLE
Port more Regex doc samples to tests

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @danmosemsft, @eerhardt, @ViktorHofer

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs [hunk 1]
```diff
@@ -137,13 +142,759 @@ string DomainMapper(Match match)
...
+                    Group group = match.Groups[groupCtr];
+                    actual.AppendLine($"   Group {groupCtr}: {group.Value}");
+                    for (int captureCtr = 0; captureCtr < group.Captures.Count; captureCtr++)
+                    {
+                        actual.AppendLine($"      Capture {captureCtr}: {group.Captures[captureCtr].Value}");
+                    }
+                }
+            }
+
+            string expected =
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: Might be easier to read to use \n in the strings, then .Replace at the end to make Environment.Newline if necessary.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs [hunk 2]
```diff
@@ -137,13 +142,759 @@ string DomainMapper(Match match)
...
+            Assert.False(Regex.IsMatch("me.myself!", Pattern, RegexOptions.IgnoreCase | options));
+        }
+
+        // https://docs.microsoft.com/en-us/dotnet/standard/base-types/miscellaneous-constructs-in-regular-expressions#inline-options
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_InlineOptions(RegexOptions options)
+        {
+            const string Input = "double dare double Double a Drooling dog The Dreaded Deep";
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: Some of these are pretty wacky!!

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: Yeah.  I refrained from applying any editorial oversight :)

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs [hunk 3]
```diff
@@ -137,13 +142,759 @@ string DomainMapper(Match match)
...
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_EngineCapabilities_ConditionalEvaluation(RegexOptions options)
+        {
+            const string Pattern = @"^(?<Pvt>\<PRIVATE\>\s)?(?(Pvt)((\w+\p{P}?\s)+)|((\w+\p{P}?\s)+))\r?$";
+
+            string input = "<PRIVATE> This is not for public consumption." + Environment.NewLine +
+                           "But this is for public consumption." + Environment.NewLine +
+                           "<PRIVATE> Again, this is confidential.\n";
```
Comment (ViktorHofer) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: Can we avoid keywords like confidential? Credscan is running in the official builds and fails when it fines suspicious phrases that might leak a token.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: Ok

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs [hunk 4]
```diff
@@ -137,13 +142,759 @@ string DomainMapper(Match match)
...
+        // https://docs.microsoft.com/en-us/dotnet/standard/base-types/backtracking-in-regular-expressions#lookahead-assertions
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_Backtracking_LookaheadAssertions(RegexOptions options)
+        {
+            const string Input = "aaaaaaaaaaaaaaaaaaaaaa.";
+
+            //const string Pattern = @"^(([A-Z]\w*)+\.)*[A-Z]\w*$";
+            //Assert.False(Regex.IsMatch(Input, Pattern, RegexOptions.IgnoreCase)); // takes too long due to backtracking
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: > // takes too long due to backtracking

That's interesting. Is this a candidate for optimization (rewriting the tree)? After that it seems we have a test ready to protect the optimization.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: > Is this a candidate for optimization (rewriting the tree)?

Into what?

Comment (danmosemsft) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]: I hadn't figured that out yet -- was thinking 😄

**Pull request merged**